Initial example of CEUS, RBSA, and feeder pipelines

dchassin opened the pull request:
This PR is provided to illustrate how we would implement the full pipelines given the design.

The pipeline is implemented in `LoadInsight.py`. It uses the example modules, functions, and data object handling. It support data-oriented check and plot (see `pipeline.add_data`) and linear tasks (see `pipeline.add_task`).

The pipeline in `selftest.py` is a simple test run to ensure everything works ok.

### Work deferred to next PR 
(See branch `example_2`)
- Make caches private to pipelines so data object names don't collide unless they are shared.
- Implement dependency tests on pipelines inputs and outputs
- Implement concurrency tests on pipelines tasks

jongoncalves commented on source/selftest.py:
```diff
@@ -0,0 +1,45 @@
+        # transfer the data to the remote storage
+        p.save()
+
+    finally:
```
Please don't ever silence an error - this is big red flag! Catch the exception - log it, handle it/ignore it, then move on. Doing things like these will break future implementation of auto crash reporting - giving us false positives.

dchassin replied:
I don't understand. Exceptions are passed through (I know this because I got plenty before I got it working). Finally just makes sure the cleanup is done regardless of whether an exception is thrown.

jongoncalves replied:
What I'm suggesting is that when this is running in the cloud, the exception doesn't get logged anywhere we can later find - all we'll know is that the pipeline stopped.

jongoncalves commented on source/LoadInsight.py:
```diff
@@ -0,0 +1,85 @@
+    external.building_mix(input="COMMERCIAL_MIX",output="commercial_mix"),
+    compose.buildings(input=["commercial_mix","commercial_enduse_components"], output="commercial_components"),
+    external.feeders(input="FEEDER_MIX", output="feeder_mix"),
+    compose.feeder(input=["commercial_components","residential_components","feeder_mix"], output="feeder_composition"),
```
Inputs should always respect a single data type (in most cases), there are exceptions of course. I'd make them all arrays. 

Better yet, I think we can provide a dict (map) of input and outputs

dchassin replied:
I haven't implemented the new format in the full LoadInsight.py yet.  I'm working on it right now.  The update is only for the example.

dchassin replied:
We could just use kwargs.

dchassin replied:
Actually we should be careful about the input and output args.  They are ordered because some of the functions may not commutative, etc.

jongoncalves commented:
Also, why are there two branches that are very similar? This is confusing and making it harder to converge into a common source of code truth (master)

dchassin commented:
Not sure. I tried to rename the old branch but I must not have done it right.

dchassin commented:
Both the selftest and LoadInsight itself are working now.

I fixed the data class so it's easier to access the underlying DataFrame.
I added support for context information in verbose and warning messages
The LoadInsight RBSA pipeline runs sequentially now (although the functions are stubbed in right now)
Current issues:

CEUS and feeder pipelines are not done yet.
dependencies and concurrency in the pipeline don't work ye

dchassin commented:
@jongoncalves: do you want me to wait until you merge to finish the todo list?

dchassin commented:
The read and write list call mapping is not implemented properly.

dchassin commented:
I fixed the run calls so they don't need to read inputs or write outputs.  Checks are still done in the wrong place though.  They need to be done by the data objects, not by the tasks.

dchassin commented:
Plotting is now added.  The main doc that explain how to use all this is in pipeline.

dchassin commented:
@jongoncalves: the work on this PR is done and I've changed it from "draft" to "review".  I will continue work on the open issues (separate caches, dependencies, and parallelism) on the next branch/PR.

@berkserbet: you can start implementing the RBSA pipeline.  You will need to do the following:
- Write the `run()` transformations in the modules that implement tasks (e.g., `clean`, `external`, `extract`, `group`,`index`,`electrification`,`normalize`,`composition`,`sensitivity`,`project`).  I may have missed some on this list and/or you may need to add any I forgot. See `LoadInsight.py` for the complete sequence.  You can also use `selftest.py` as a guide for how to add checks and plots to data objects.

- Write the `check.py` functions for data objects
- Write the `plot.py` functions to plot data objects.

jongoncalves commented:
I'm starting to bring this in to master - but I won't be able to merge it straight in bc it's quite different structurally then master's setup. Also, using `selftest.py` is not the way to go - this should be done via unit tests - we don't need re-invent the wheel here.

jongoncalves commented on source/pipeline.py:
```diff
@@ -0,0 +1,84 @@
+        self.datalist = {}
+        for data in datalist:
+            self.add_data(data)
+        return self.datalist
```
this convention of adding setters and getters is odd - especially since the setter (and the adder above) is returning a value. This seems atypical or semantically confusing.

dchassin commented:
What the guidance I can follow on this? I can reorg easily enough.

D.

On Jun 8, 2019, at 07:49, Jonathan Goncalves <notifications@github.com<mailto:notifications@github.com>> wrote:


Also, why are there two branches that are very similar? This is confusing and making it harder to converge into a common source of code truth (master)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/slacgismo/loadinsight/pull/4?email_source=notifications&email_token=AE6WRR5WT7BRIJ5YTTA6FNTPZPBHDA5CNFSM4HVXQA22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXHWA7I#issuecomment-500129917>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE6WRR7CW5DD45FRHRNEQETPZPBHDANCNFSM4HVXQA2Q>.

dchassin commented:
How would suggested naming them?

On Jun 9, 2019, at 13:55, Jonathan Goncalves <notifications@github.com<mailto:notifications@github.com>> wrote:


@jongoncalves commented on this pull request.

________________________________

In source/pipeline.py<https://github.com/slacgismo/loadinsight/pull/4#discussion_r291850083>:

> +        """Add a task to the pipeline"""
+        self.tasklist.append(entry)
+
+    def set_tasks(self,tasklist = []):
+        self.tasklist = tasklist
+
+    def add_data(self,**kwargs):
+        item = data(**kwargs)
+        self.datalist[kwargs["name"]] = item
+        return item
+
+    def set_data(self,datalist = []):
+        self.datalist = {}
+        for data in datalist:
+            self.add_data(data)
+        return self.datalist


this convention of adding setters and getters is odd - especially since the setter (and the adder above) is returning a value. This seems atypical or semantically confusing.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/slacgismo/loadinsight/pull/4?email_source=notifications&email_token=AE6WRR4WJRSNV2Q3Q5YWGQ3PZVU4XA5CNFSM4HVXQA22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB27T55A#pullrequestreview-247414516>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE6WRR2FVY7MWJRUAK6RRB3PZVU4XANCNFSM4HVXQA2Q>.

dchassin commented:
1. Changed `selftest.py` to `example.py`. 
2. Renamed `add_data` to `create_data`.  
3. Removed return value from `set_data`.

jongoncalves commented on source/extract.py:
```diff
@@ -0,0 +1,27 @@
+
+config = load_config()
+
+class rbsa :
```
There's another `class rbsa` in a file called `clean.py`, is this accidental? Their definition is identical, so I assume so...

dchassin commented:
They will differ eventually.  They are stubs now. One is rbsa extract, the other is rbsa clean.

On Jun 9, 2019, at 19:56, Jonathan Goncalves <notifications@github.com<mailto:notifications@github.com>> wrote:


@jongoncalves commented on this pull request.

________________________________

In source/extract.py<https://github.com/slacgismo/loadinsight/pull/4#discussion_r291867053>:

> @@ -0,0 +1,27 @@
+from utils import *
+
+config = load_config()
+
+class rbsa :


There's another class rbsa in a file called clean.py, is this accidental? Their definition is identical, so I assume so...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/slacgismo/loadinsight/pull/4?email_source=notifications&email_token=AE6WRR6AZBASWG664JOW7F3PZW7D7A5CNFSM4HVXQA22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB27ZAEQ#pullrequestreview-247435282>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE6WRR247XXQ6734TKAFHWLPZW7D7ANCNFSM4HVXQA2Q>.

dchassin commented:
Can we close this without merging?

jongoncalves commented:
> Can we close this without merging?

Yep
